src/main/scala/io/iohk/ethereum/merklepatriciatrie/MerklePatriciaTrie.scala

```diff
@@ -0,0 +1,486 @@
+          case ml =>
+            // Partially shared prefix, we have to replace the node with an extension with the shared prefix
+            val (sharedKeyPrefix, sharedKeySuffix) = sharedKey.splitAt(ml)
+            val temporalExtensionNode = ExtensionNode(sharedKeySuffix, next, hashFn)
+            val NodeInsertResult(newBranchNode: BranchNode, toDeleteFromStorage, toUpdateInStorage) = put(temporalExtensionNode, searchKey.drop(ml), value)
+            val newExtNode = ExtensionNode(sharedKeyPrefix, newBranchNode, hashFn)
+            NodeInsertResult(
+              newNode = newExtNode,
+              toDeleteFromStorage = node +: toDeleteFromStorage,
+              toUpdateInStorage = newExtNode +: toUpdateInStorage
+            )
+        }
+      case branchNode@BranchNode(children, optStoredValue, _) =>
+        if (searchKey.isEmpty) {
+          // The key is empty, the branch node should now be a terminator node with the new value asociated with it
+          val newBranchNode = BranchNode(children, Some(value), hashFn)
+          NodeInsertResult(
+            newNode = newBranchNode,
+            toDeleteFromStorage = Seq(node),
+            toUpdateInStorage = Seq(newBranchNode)
+          )
+        }
+        else {
+          // Non empty key, we need to insert the value in the correct branch node's child
+          val searchKeyHead: Int = searchKey(0)
+          val searchKeyRemaining = searchKey.tail
+          if (children(searchKeyHead).isDefined) {
+            // The associated child is not empty, we recursively insert in that child
+            val NodeInsertResult(changedChild, toDeleteFromStorage, toUpdateInStorage) =
+            put(getChild(branchNode, searchKeyHead, dataSource).get, searchKeyRemaining, value)
+            val newBranchNode = branchNode.updateChild(searchKeyHead, changedChild, hashFn)
+            NodeInsertResult(
+              newNode = newBranchNode,
+              toDeleteFromStorage = node +: toDeleteFromStorage,
+              toUpdateInStorage = newBranchNode +: toUpdateInStorage
+            )
+          }
+          else {
+            // The associated child is empty, we just replace it with a leaf
+            val newLeafNode = LeafNode(searchKeyRemaining, value, hashFn)
+            val newBranchNode = branchNode.updateChild(searchKeyHead, newLeafNode, hashFn)
+            NodeInsertResult(
+              newNode = newBranchNode,
+              toDeleteFromStorage = Seq(node),
+              toUpdateInStorage = Seq(newLeafNode, newBranchNode)
+            )
+          }
+        }
+    }
+  }
+
+  private def remove(node: Node, searchKey: Array[Byte]): NodeRemoveResult = (node, searchKey.isEmpty) match {
+    // They key matches a branch node but it's value doesn't match the key
+    case (BranchNode(children, None, _), true) => NodeRemoveResult(hasChanged = false, maybeNewChild = None)
+    // We want to delete Branch node value
+    case (BranchNode(children, _, _), true) =>
```

Comment:
@adamsmo as in that case we are removing branch node value  `BranchNode(children, None, hashFn)` (check the _None_ as second parameter) you need to check if you have a single node within your 16 positions, if so, that's an invalid state (Branch node where there is only a single entry). Check out this case https://github.com/input-output-hk/etc-client/blob/feature/patriciaTrie/src/main/scala/io/iohk/ethereum/merklepatriciatrie/MerklePatriciaTrie.scala#L450